pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [INSERT statement too large]

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkload [failed](https://tanzanite.cluster.engflow.com/invocations/default/6a723732-b1cb-4f15-85f0-4a6c68bc048d?testReportRun=25&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvd29ya2xvYWQvc2NoZW1hY2hhbmdlOnNjaGVtYWNoYW5nZV90ZXN0) on release-25.1 @ [6123ebaa10c3847899aa81b09ec4cb8d08c0bd6d](https://github.com/cockroachdb/cockroach/commits/6123ebaa10c3847899aa81b09ec4cb8d08c0bd6d):


```
     ]
    ]
   }
  ],
  "previousStatements": [
   "INSERT INTO public.table_w0_43 (col43_w0_44, col43_w0_45, col43_w0_46, col_w0_43_w1_142, col43_w0_48, col43_w0_49, col43_w0_50, col43_w0_51, col43_w0_52, col43_w0_53) VALUES (e'\\x12'::BPCHAR,'01070000800400000001020000800300000068DD1C5F83A1D941CB3AC71D9B42F9C1160E5EC150DEE8C1B8697C8F89C7DDC10E18C26A8DD7EBC1E0B2F24B84A8CF4132BEDE51587801C2603768669BA0EC41D060345D3DE9DD410106000080090000000103000080010000000D0000007EFAD05F7A51F9C1A89FA8D8F995F2C1761ED9857A8DF841E6ADD0629096F8C1417B7219DAD7F7C1C44208728B06DFC1F836BB4724B5E74162667F0DD8E1FBC19C6FC037167FDCC1A8C8B3481679FB416FDF0FB9410F02C208BAF021930FF0415C5E65522945FC41C160AE856250F4C12079AF470A87EC419625B3C444CDF741481119486A73D9C182975851773DF441FE1F6FC05EB8F6413079B67EFCF4EB410E1EFEBF8183F6C1046A6EAB1BE8E541F8530FF7F6C8F64154A074B1C36FE2412781EC3891ABF0C1F42ECF91F484FE4122277578A7EEF6C1F9096F59612BF5C1A0867F8CD7FA0142182C8F01E957E141FA9F1DDCE593F4C14C8C4DFE3A940042F4CEC73250C40042AC5C96ADCED1E5C1E866AF03115CDD41A08A2C3EFA5BFFC17EFAD05F7A51F9C1A89FA8D8F995F2C1761ED9857A8DF841010300008001000000080000002AAE99FB282AFAC11AF8F27C2EA901C2CCF368FC779DE84188D2E664EF2DE841F9603ECC44C4FCC126DB7D298B3DE9C1F48337BC357AFC4110E7A55483FFD641F0AFBD5CBB81D641C0741223904EFB41408BCF289D2AE8411442D681854EECC10C54003871F8F5412877D104A324F741E053E4651BFFF4C178BEBD4B1C94F54106EC8A9EA4B2F9411907FEF00F08F4C1A42B113BC875F6C1FC3A3E873CCFEA41E906B19A098402C22AAE99FB282AFAC11AF8F27C2EA901C2CCF368FC779DE8410103000080010000000C00000007E311534D6F02C28653BB5C1114F8C18CD8431EE97EE441A56373FA8724F4C1AC428E7E63FBFAC1760DEE04914DF94106DF305D58E9E2C10041BD4AD523FFC1B0ED798E7375D24148C99F9C52B0FA41D6B9D866E45DFEC1146CB4A6B04DE1C1FC83A065B03DE64180246F00CEC0B6413013DBF288AEC141C0D9FB96E51FCAC1605FD371D331CC41E07148489E8CB8C1680BE375D352EB41886E689F24780242F60C1FAD684DF7C17E00061743F0E1C136319AA3C4B1F341046909688110E8416F40128E93F1F8C1F0EFBEE8F09BF841701B4767F504FD4182A5AD049472F0C19C3B48EED86DE641E224A1D640B2FAC10ABEF369D2E401C2BC84665E6560F241E60679E4BB7BFAC107E311534D6F02C28653BB5C1114F8C18CD8431EE97EE4410103000080010000000C00000060593BD2EED9DAC1B80BF8F51CF7D441DC8534458055E8412062665DB368C541BD0426BA61DBF4C1ADAE429A98B0F3C11A024CA0637EF14154E714296856E4C1293C687BCD7301C2F063F3EAF97DEE4140995691323DC7C1E0F208553CD4B34104CCE0DF2303EC41F8D0BF14F29EDA41702239C8593DCF41BE8ADBCF6692F641ACE072DD5E98EA41D850C2D5AADBEDC1409109ACEDA2DCC1C85C5C48C83D0242F4FC50C162C9FCC14C0B4875F762E5C1D0D6D299A3E7FE4198B78454ED64E1C184851B7EB872DEC11AF39355E34DF841E6BEFC9B9752F84136F6845B9BE4F9C140866E73EC20FF4160B0E579E22EF941127DAB5D1E49F1C1388D2742685FEB416086F43C8F21F44160593BD2EED9DAC1B80BF8F51CF7D441DC8534458055E8410103000080010000000C000000DC5E1BFF0D89FCC17C750C3FB043F6C1C0C6DBCBAAD3FEC180505B87095CAAC1709205109DEDCBC18CB1C2A981EEE9C1283AFEF17AC9D0410089BB303FE18BC1D099C0DFAFE4F7413C0CF6E180E1ED41E8C9639B1BC1F7C180EF07CB9EBD9CC1344DDFADFA130042F8B15CC72915FCC178F1257FCB62F3419843B4149770E741CCAB3A51F3F6E541E66339D0407FF9C1D615BD4F9944F841B0B45A39B907FE411E03A9C63D1BF541402BFE3D2225F3415CB1F111BF88014290D6C86A7D06D241B08D8D5D2653DA4130CFCC53B335F44150DF1A28C968C6C15AE0588CD1A3E3C160042B5B83EFE741184FA6CEA3DBE74100AF9F96FB10E5C160315F300908D74109D28805519CFBC1DC5E1BFF0D89FCC17C750C3FB043F6C1C0C6DBCBAAD3FEC101030000800100000005000000D2793F3E8640F141D4A1BA3B9504E8417B8844EA9329F2C1C05A10F63C93E441BA00D604ABD5F64124B4D658CAA50042F84D33843773E341568D11F3850A00425A8555FF5B4E0042A0B37241B03FD3C1386194C854E9FD41845353F6F302E541D2793F3E8640F141D4A1BA3B9504E8417B8844EA9329F2C10103000080010000000A000000B443BBF1630501C2C20C23622C40F5C1E4F7D59154ACEAC11014D3A97963C441996AD48D5E2602C21E9B567CE6EA0142A64BDBCA642CF34180248263A7C6FAC1CA73E1716315F24118ADDEDD73A7E74158CF37A3C59AE041D24157669F6DF541DC92B3914796E74108F5BD672688FB4130646805E65FF341004764A499D79641983DDEACEB130242E024C61DE501E841383F19B62FDCD0C1BC026EF25F78FD41806886C37778C441E62AD3C3730DE1C1003D5E0914D2E441B0B138FE793CC2C12FF0A13B04DD00C2FC4C3D9517B4F841B4F4A40AEF91F041B443BBF1630501C2C20C23622C40F5C1E4F7D59154ACEAC101030000800100000005000000008E83E873B79641A60EEB03AD4FF3C1586A4A9BC042D0415876056050D5DE4163A0685E016DF5C12C384DED9F2CE2C17608E6BD1398F14150B0048AFD29FA41D4E9DB6FEE7EF341304378B07957CBC13033B05E2935D1C100ABFF134B09F941008E83E873B79641A60EEB03AD4FF3C1586A4A9BC042D0410103000080010000000B000000ED3AFA061E6AFEC1206A729EC49BC741E8E6D016F96EEE41B6A3223C2F78E3C130745CEABCE6C1C15E5737CF43D9F1C1409637CD5ED4C04118E8E402C03BF0C1B8C8C64298BFFF412070C18306F0CF4130A54742B1A4D7C106CD336F0CF1F3C1641A4BDAB8E5F341849233BD176AED41AEC97DB2EE61E7C108049AFCCFDFF3412467CF613ABCFE4130F27D6C6672D3C168CE229723AEEE41147023A50A06FC41F803C4E71A62F141A44F2652967EF241443E3306795C0242ACBDE2B2C7230242A3A1928A80C601C236D7115077A0F041D09651F2EF74FD41D4D8812F236001C254657EE7BA57E941C8D9FB2070B9FF41ED3AFA061E6AFEC1206A729EC49BC741E8E6D016F96EEE41010600008007000000010300008001000000070000004AE0BC17B20FFCC15A8AA4756CC5FEC1283EE9E0C1A2EDC140F2133359FCDD41E739D71908B8FFC1B04B4DD991B4EA4110D1D59DC4DADF419450F9A7174EEC41A079D0CA1FFACA41489512155B71D6C1B84C520CD620E6419EF90834CA22F841B7A7F777F73802C2DAED682150E9F14178CD5E084EE40142207E750D877D02C218DE50A61953D841BCADA1F42448DFC14AE0BC17B20FFCC15A8AA4756CC5FEC1283EE9E0C1A2EDC101030000800100000008000000580ED5C4B1BE01C26B0449DFAE52FEC1A456463D7EB0E541A42642A57152EDC1D0CE568BF554CDC16E10A928A882FDC11C6C7460093BE9C12CAC41F52444EEC1B04A0AF052C4E341FE30E0DAD6A9F441286FFAF33507EE4101F1667E1F73F7C1B0F4C091D793CDC1A0956C2992B2D6418C8B3FC40306D9C1846BA0F997D9EFC1E4F87B1AAE70FC411CF5BB35FA4CF24112E3102B9B1EFAC180438C46E190E74103357937F74002C2580ED5C4B1BE01C26B0449DFAE52FEC1A456463D7EB0E541010300008001000000060000002DB5F6EF652BF1C1B7479EA84A12F2C195F81AB6F694F5C1006671DEC515C4C14CE1EA1CC4E2EBC16615F72835E3EAC17C4F320C15CEEE41914A537D8A93F1C1528426D3A49FE1C14087C8D7B3B2F841C0F493C80D3BA841CCCE15082789F9416CA2CDE6F637FC41A0CCD3CD2A3CF541B8CD39C724F4DBC12DB5F6EF652BF1C1B7479EA84A12F2C195F81AB6F694F5C101030000800100000008000000D521B088240AF7C1E165DB8C45A6FEC13064215546D5CB41421F0966F44BF24160A969906B4DF6C118DFD19619ADD94190061C30D6910242C9204C06661202C2A59C67B8B9F7F3C1ACA82C22E8BBF541DDF4A5751C17F1C176F2905ED72EF441188D99A794A4FE4180834E97BC1A9CC16C05F22E6677014246F7FD97D3F5EDC1A67BEE991E5C0242A0F1E045ED08F641A7658B4EF502F1C168B94150772BF241D8C58A673ADCD541D521B088240AF7C1E165DB8C45A6FEC13064215546D5CB410103000080010000000A000000CC0B32F016F4F0C10C58AD2C8D27D5C1D9F3E1B24B9A01C248C76519AF4FD0C1ECCA650AF933E9C16EE1D6715888FBC1DA0CBC2658A8F84120C139613C77C5C1DEE617D26DBCF641768AD11B331DF041607748D29AF9C041B4D26CF8D8D6D7C17A63B1725C49F141A6C4491835690042522B5AD51954F1410C87F447EF67D9C1D03D7D19E7670242B8E7F463ECDBDF417A30786D17B400C2D8EE181F92770142F0B53514B6E6E341BAC48A6F59DFFEC1D4440ECD8F8AF841B23E9186CD27F5C1E129248BE29AFAC1C29037BDC50CF041CCFF1BA73487FD41CC0B32F016F4F0C10C58AD2C8D27D5C1D9F3E1B24B9A01C20103000080010000000500000057DD5298673FF0C1903D657777E3DA414C2707EE4E60E7415055C315F98DC041445D929957E9E041542B5BA6F3B6E4C18EC79EFB59B9F04158CBD41A9FB9E3C114FDC083F045E84106C9F9E44947F141AE21C7748A0D02423403C48A566BF34157DD5298673FF0C1903D657777E3DA414C2707EE4E60E7410103000080010000000B000000F883A5FA2E3DD341D098F93B75C9FDC1D69D5173470FFBC1841432416317E941BAE3E84197D0F3C1B49D9DAB3DAA0042F694733F02C8F0414B6247756393F2C140A03F8761F2CCC1B463FCA5B369FE41D2A204031637E6C1808C20DD47D4AD41D0412060FC48FE4128ABA9FDB3C3D0C127FF1F231B49F6C1BE8D99D18768F841706CB3764112CAC196C3D6BB4531FCC14C77DE1FFFA0FE41405B13D49832A4C1F08E165E229EE8416CB623EE6608F34120103BE5AFF3B4C1BC214F5468A5FD4190A7CCCAFB3CFE419CF226333491FD41ECB32158CC180142A321D17B018A00C250D89F3BAD38CEC109B96BB95E3EFFC1F883A5FA2E3DD341D098F93B75C9FDC1D69D5173470FFBC10101000080FCE36FCA0D8EF0C158E382E3015EE5C1BA3BD1B78C0FF441':::GEOMETRY,'0001-01-01 00:00:00':::TIMESTAMP,3614:::OID,90000:::OID,(-26):::INT8,'''XlW'' ''ccyMxvoITk''':::TSVECTOR,NULL,3614:::OID,e'\\U000A1327\\U000E28B2\\U000DCBF8\\U0000A385\\U0009467B' COLLATE en_US),('C'::BPCHAR,'010500000003000000010200000009000000C2DF54770A6CF741C41AF94F3AD4EEC10E531C156563F9410E77813ECEFAEDC13E8582FB0341004249C166FDFBD1F4C1667840168642F94110202E2DD4C0C7C180D14635A0DEFF41E0A4872E6875FD4146B25831BABD00427875CA886A180242D6386C6AD214E6C1A86E54ECB195F04153D8B23D43FC01C2F87FF3EDB092F941B4F024850E36E3C140EADC555B5FEE41010200000005000000A6B9D9DEFEFDF541DA46AE6B2376EEC1A47F95426604F941803C309157CBFA4100578D533CEEAFC18CC9126C792EFC4186164EC46058E8C1029A2F97D7F9014204273C6980DCF7C1841A0C1D715AFE41010200000002000000026782833B7DFCC11876483D8815EC41AC00FF2853D6FFC1C4249A1CF104E841':::GEOMETRY,'2015-01-08 16:33:40.000174':::TIMESTAMP,26:::OID,25:::OID,108:::INT8,'''OEeFyq'':468A,622B,840C,961C ''POM'' ''WpBcL'':8C,99A,333C ''WycobQyRE'':204B,919A,980C ''fUWjcYtnGo'' ''jrRBbjlO'':234A ''vSvjJAE'':73C,309C ''vY'' ''vaqL'':948C':::TSVECTOR,true,90002:::OID,e'\\U00053037\\U000366D4\\U000625ED' COLLATE en_US)"
  ],
  "usesDeclarativeSchemaChanger": true
 }
}
    schema_change_external_test.go:131: 
        	Error Trace:	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:131
        	Error:      	Received unexpected error:
        	            	***UNEXPECTED ERROR; Received an unexpected execution error.: ERROR: command is too large: 67674915 bytes (max: 67108864) (SQLSTATE XXUUU)
        	            	(1) forced error mark
        	            	  | "txn needs to rollback"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2) forced error mark
        	            	  | "fatal error when running txn"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (3)
        	            	Wraps: (4) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*opStmt).executeStmt
        	            	  | 	pkg/workload/schemachange/operation_generator.go:3215
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).runInTxn
        	            	  | 	pkg/workload/schemachange/schemachange.go:511
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
        	            	  | 	pkg/workload/schemachange/schemachange.go:608
        	            	  | pkg/ccl/testccl/workload/schemachange/schemachange_test_test.TestWorkload.TestWorkload.func4.func6
        	            	  | 	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:120
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	external/org_golang_x_sync/errgroup/errgroup.go:78
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1695
        	            	Wraps: (5) ***UNEXPECTED ERROR; Received an unexpected execution error.
        	            	Wraps: (6) ERROR: command is too large: 67674915 bytes (max: 67108864) (SQLSTATE XXUUU)
        	            	Error types: (1) *markers.withMark (2) *markers.withMark (3) *schemachange.ErrorState (4) *withstack.withStack (5) *errutil.withPrefix (6) *pgconn.PgError
        	Test:       	TestWorkload
Schema Workload Stats
Total Schema Statements Executed = 93
Total Schema Statements Succeeded = 93
Total Schema Statement Expected Failures = 0
Total Transactions Committed = 66
Total Transactions Rolled Back = 53
Total Transactions Executed = 119
    schema_change_external_test.go:96: backup, tracing data, and system table dumps in /var/lib/engflow/worker/work/3/exec/bazel-out/k8-fastbuild/testlogs/pkg/ccl/testccl/workload/schemachange/schemachange_test/run_25_of_25/test.outputs/logTestWorkload4176391018
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestWorkload4176391018
--- FAIL: TestWorkload (17.55s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=25</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #139331 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [duplicate constraint name] [C-test-failure O-robot T-sql-foundations branch-release-24.2]
- #138721 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [C-test-failure O-robot P-3 T-sql-foundations branch-release-23.1]
- #138547 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [duplicate constraint name] [C-test-failure O-robot P-2 T-sql-foundations branch-release-24.1]
- #137972 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [requested column not in index] [C-test-failure O-robot P-2 T-sql-foundations branch-release-24.3.3-rc]
- #136966 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [duplicate key value violates unique constraint] [C-test-failure O-robot P-2 T-sql-foundations branch-master]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkload.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-46686

--- a/pkg/workload/schemachange/operation_generator.go
+++ b/pkg/workload/schemachange/operation_generator.go
@@ -2935,7 +2935,23 @@ func (og *operationGenerator) insertRow(ctx context.Context, tx pgx.Tx) (stmt *o
 	for i := 0; i < numRows; i++ {
 		var row []string
 		for _, col := range nonGeneratedCols {
-			d := randgen.RandDatum(og.params.rng, col.typ, col.nullable)
+			// Limit the size of columns being generated.
+			const maxSize = 1024 * 1024
+			maxAttempts := 32
+			var d tree.Datum
+			for i := 0; i < maxAttempts; i++ {
+				d = randgen.RandDatum(og.params.rng, col.typ, col.nullable)
+				// Retry if we exceed the maximum size.
+				if d.Size() < maxSize {
+					break
+				}
+			}
+			if d.Size() > maxSize {
+				og.LogMessage(fmt.Sprintf("datum of type %s exceeds size limit (%d / %d)",
+					col.typ.SQLString(),
+					d.Size(),
+					maxSize))
+			}
 			// Unfortunately, RandDatum for OIDs only selects random values, which will
 			// always fail validation. So, for OIDs we will select a random known type
 			// instead.

